src/lightning/pytorch/loops/utilities.py

```diff
@@ -124,16 +124,25 @@ def _reset_progress(loop: _Loop) -> None:
 
 
 def _set_sampler_epoch(dataloader: Iterable, epoch: int) -> None:
-    """Calls the ``set_epoch`` method on either the sampler or the batch sampler of the given dataloader.
+    """Calls the ``set_epoch`` method on either the sampler of the given dataloader.
 
-    Every PyTorch dataloader has either a sampler or a batch sampler, and if it is wrapped by a
+    Every PyTorch dataloader has either a sampler or a batch sampler. If the sampler is wrapped by a
     :class:`~torch.utils.data.distributed.DistributedSampler`, ``set_epoch`` must be called at the beginning
     of every epoch to ensure shuffling applies a new ordering. This has no effect if shuffling is off.
     """
-    for sampler_name in ("sampler", "batch_sampler"):
-        sampler = getattr(dataloader, sampler_name, None)
-        if sampler is not None and callable(getattr(sampler, "set_epoch", None)):
-            sampler.set_epoch(epoch)
+    objects = set()
+    # check dataloader.sampler
+    if (sampler := getattr(dataloader, "sampler", None)) is not None:
+        objects.add(sampler)
+    # check dataloader.batch_sampler.sampler
+    if (batch_sampler := getattr(dataloader, "batch_sampler", None)) is not None and (
+        sampler := getattr(batch_sampler, "sampler", None)
+    ) is not None:
+        objects.add(sampler)
+    for obj in objects:
+        set_epoch = getattr(obj, "set_epoch", None)
+        if callable(set_epoch):
+            set_epoch(epoch)
```

Comment:
I think you also need to add dataloader.batch_sampler to the list. What if the batch sampler = DistributedSampler? This is possible if the user wraps their (custom) batch sampler like that.

Reply:
Can you show an example of this? `DistributedSampler` takes a dataset

If I try it, I get:

```python
Traceback (most recent call last):
  File "/home/carmocca/git/lightning/examples/pl_bug_report/bug_report_model.py", line 70, in <module>
    run()
  File "/home/carmocca/git/lightning/examples/pl_bug_report/bug_report_model.py", line 67, in run
    trainer.fit(model)
  File "/home/carmocca/git/lightning/src/lightning/pytorch/trainer/trainer.py", line 425, in fit
    call._call_and_handle_interrupt(
  File "/home/carmocca/git/lightning/src/lightning/pytorch/trainer/call.py", line 42, in _call_and_handle_interrupt
    return trainer.strategy.launcher.launch(trainer_fn, *args, trainer=trainer, **kwargs)
  File "/home/carmocca/git/lightning/src/lightning/pytorch/strategies/launchers/subprocess_script.py", line 92, in launch
    return function(*args, **kwargs)
  File "/home/carmocca/git/lightning/src/lightning/pytorch/trainer/trainer.py", line 464, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/home/carmocca/git/lightning/src/lightning/pytorch/trainer/trainer.py", line 836, in _run
    results = self._run_stage()
  File "/home/carmocca/git/lightning/src/lightning/pytorch/trainer/trainer.py", line 875, in _run_stage
    self._run_train()
  File "/home/carmocca/git/lightning/src/lightning/pytorch/trainer/trainer.py", line 896, in _run_train
    self.fit_loop.run()
  File "/home/carmocca/git/lightning/src/lightning/pytorch/loops/fit_loop.py", line 176, in run
    self.setup_data()
  File "/home/carmocca/git/lightning/src/lightning/pytorch/loops/fit_loop.py", line 217, in setup_data
    dl = trainer._data_connector._prepare_dataloader(dl, shuffle=shuffle, mode=RunningStage.TRAINING)
  File "/home/carmocca/git/lightning/src/lightning/pytorch/trainer/connectors/data_connector.py", line 257, in _prepare_dataloader
    dataloader = _update_dataloader(dataloader, sampler, mode=mode)
  File "/home/carmocca/git/lightning/src/lightning/pytorch/utilities/data.py", line 140, in _update_dataloader
    dl_args, dl_kwargs = _get_dataloader_init_args_and_kwargs(dataloader, sampler, mode)
  File "/home/carmocca/git/lightning/src/lightning/pytorch/utilities/data.py", line 201, in _get_dataloader_init_args_and_kwargs
    dl_kwargs.update(_dataloader_init_kwargs_resolve_sampler(dataloader, sampler, mode, disallow_batch_sampler))
  File "/home/carmocca/git/lightning/src/lightning/pytorch/utilities/data.py", line 305, in _dataloader_init_kwargs_resolve_sampler
    batch_size=batch_sampler.batch_size,
AttributeError: 'DistributedSampler' object has no attribute 'batch_size'
```

Reply:
We only require this `.batch_size` attribute exists. I think that's a fair constraint to have, but if you think otherwise, opening an issue about it is good with me. Maybe request for this should come from the community though

Reply:
I was just confused. It seems we had batch_sampler.set_epoch defined on our wrapper, that's why we called it. It just looked like we are removing support for that.